pkg/apis/workflow/v1alpha1/workflow_types.go

```diff
@@ -3766,13 +3773,26 @@ func (ss *SemaphoreStatus) LockWaiting(holderKey, lockKey string, currentHolders
 	return true
 }
 
+func GetHolderKeyVersion() HoldingNameVersion {
+	switch os.Getenv("HOLDER_KEY_VERSION") {
+	case "v2":
+		return HoldingNameV2
+	default:
+		return HoldingNameV1
+	}
+}
+func GetHoldingNameV2OrV1(holderKey string) string {
+	switch GetHolderKeyVersion() {
```

Comment:
Is the env var actually necessary? Since you have `CheckHolderKeyVersion`, this seems like you should always be able to use V2 unless V1 was already used

Reply:
yes it is, it still is needed in LockAcquired calls, you are right about it not being needed for the Released calls.

Reply:
For `LockAcquired`, you can just always use V2, no?

Reply:
Yeah you can if you use `CheckHolderKeyVersion` but why? It's much cleaner to just use the env var (which you do need for `LockReleased`). Why infer information when you have it?

Reply:
No we shouldn't do that, this will change how the workflow status  is populated. This should be opt in at least for sometime imo.

Reply:
It does, but it is a fix and it is objectively better than the previous option.

Unlike `POD_NAMES`, there is also little to no client-side usage of this that would incur breakage

Reply:
Alright lets wait on a tie breaking vote here, I think @Joibel is with you on this one.

Reply:
I think we should drop support for legacy holding keys, and just run with the new code.

Reasoning:
* Just V2 is simple
* There is a lot of code here supporting both key types
* Removing support for v1 (except on upgrade) reduces the amount of code even further.